src/socket.rs

```diff
@@ -158,30 +156,40 @@ impl<A: NetworkAddress, S> Socket<A, S> {
     ) -> std::io::Result<Option<Timestamp>> {
         const TIMEOUT: Duration = Duration::from_millis(200);
 
-        if let Ok(timestamp) = tokio::time::timeout(TIMEOUT, async {
-            loop {
-                self.errqueue_waiter.wait().await?;
+        match tokio::time::timeout(TIMEOUT, self.fetch_send_timestamp_loop(expected_counter)).await
+        {
+            Ok(res_opt_timestamp) => res_opt_timestamp,
+            Err(_timeout_elapsed) => Ok(None),
+        }
+    }
 
-                match self.fetch_send_timestamp_help(expected_counter) {
-                    Ok(Some(timestamp)) => return Ok(Some(timestamp)),
-                    Ok(None) => continue,
-                    Err(error) => {
-                        tracing::warn!(error = ?error, "Error fetching timestamp");
-                        return Err(error);
+    #[cfg(target_os = "linux")]
+    async fn fetch_send_timestamp_loop(
+        &self,
+        expected_counter: u32,
+    ) -> std::io::Result<Option<Timestamp>> {
+        loop {
+            // the timestamp being available triggers the error interest
+            let mut guard = self.socket.ready(Interest::ERROR).await?;
+
+            match self.fetch_send_timestamp_try_read(expected_counter) {
+                Ok(Some(timestamp)) => break Ok(Some(timestamp)),
+                Ok(None) => continue,
+                Err(error) => {
+                    if error.kind() == std::io::ErrorKind::WouldBlock {
+                        guard.clear_ready();
+                        continue;
+                    } else {
+                        tracing::warn!(error = ?&error, "Error fetching timestamp");
+                        break Err(error);
```

Comment:
Why not use the `async_io` function here? The tokio guard interface seems quite complex to me, and I think you may be using it incorrect here. (the docs say something about checking the interests which are returned as ready, which this doesn't seem to do, but I am not sure how critical that is.)

Reply:
Fixed. I think it was correct in practice (there is only one interest, the error interest, that could trigger and therefore could block). But using `async_io` is much nicer.